runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/io/GzipByteReadChannel.kt

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.io
+
+import aws.smithy.kotlin.runtime.InternalApi
+import java.util.zip.GZIPOutputStream
+
+@InternalApi
+public class GzipByteReadChannel(
+    private val channel: SdkByteReadChannel,
+) : SdkByteReadChannel by channel {
+    private val gzipBuffer = SdkBuffer()
+    private val gzipOutputStream = GZIPOutputStream(gzipBuffer.outputStream())
+
+    /**
+     * Keeps track of whether a read operation has been made on this byte read channel
+     */
+    private var read: Boolean = false
+
+    override suspend fun read(sink: SdkBuffer, limit: Long): Long {
+        require(limit >= 0L)
+        if (limit == 0L) return 0L
+
+        if (isClosedForRead) {
+            if (!read) { // Empty payload
+                gzipOutputStream.write(ByteArray(0))
+                gzipOutputStream.close()
+                gzipBuffer.readAll(sink)
+                gzipBuffer.close()
+
+                read = true
+            }
+
+            return -1
+        }
+
+        if (!read) read = true
+
+        val temp = SdkBuffer()
+        val rc = channel.read(temp, limit)
+
+        gzipOutputStream.write(temp.readByteArray())
+        gzipBuffer.readAll(sink)
```

Comment:
**correctness**: User didn't ask to read all though they asked to read up to the limit.

```kotlin
        if (rc == -1L) {
            // may trigger additional bytes written by gzip defalter
            gzipOutputStream.close()
        }

        // source is exhausted and nothing left buffered we are done
        if (rc == -1L && gzipBuffer.exhausted()) return -1L

        // compress what we read and add it to the buffer
        if (rc >= 0L) {
            gzipOutputStream.write(temp.readByteArray())
        }

        // read bytes read from compressed content
        return gzipBuffer.read(sink, limit)

```